### ISSUE TITLE
Option to not walk back cursor in Vim mode when exiting insert mode

### ISSUE THREAD

**Issue by verhovsky** [zed-industries/zed#17676]
To see the issue:

1. Enable Vim mode
2. Put cursor at the end of a long line
3. Repeatedly press `i` + Escape

Actual behavior: the cursor will eventually end up at the beginning of the line

Desired: I would like an option that would make exiting insert mode keep the cursor where it was instead of one character back. There are a few other places where this option should have an effect, like `C`.

Emac's evil mode has `(setq evil-move-cursor-back nil)` for this. VSCode Vim refused to implement this https://github.com/VSCodeVim/Vim/issues/7548

**Comment by graelo** [comment]
Hi, I'm a vim user.

If your proposal is implemented, I hope it is only an option and never the default behavior, simply because it would be a significant departure from vim's model.

Best,

**Comment by ConradIrwin** [comment]
@verhovsky You can do something like: `"escape": ["workspace::SendKeystrokes", "ctrl-c l"]` to implement this for yourself.

**Comment by verhovsky** [comment]
That doesn't work at the start of the line and it also doesn't work with paste, as I described in https://github.com/VSCodeVim/Vim/issues/7548#issuecomment-1068981825

**Comment by ConradIrwin** [comment]
@verhovsky Makes sense. I do want to fix some of this stuff (in particular being able to bind to multiple actions with one key-stroke; but we've foolishly already used array syntax for passing arguments to bindings).

How does evil mode handle `a <escape>` do you end up moving forward? It seems very odd to me!

At some point I'd love if we could have extension support that would let you do this; until then we could add a setting in vim mode, but I'm worried about adding a lot of conditional logic to emulate various different vim emulators :D.

**Comment by verhovsky** [comment]
> already used array syntax for passing arguments

I noticed that, I tried using nested arrays

```json
      "p": [["vim::Right"], ["vim::Paste"]],
```

which would work unambiguously but it's pretty meh because now if you have a config of a regular array and you add a value in an array, that would change the meaning of existing values. Like if you have

```json
      "p": ["workspace::SendKeystrokes", "x y z"]
```

and you wanted to add

```json
      "p": ["workspace::SendKeystrokes", "x y z", ["vim::Right"]]
```

it would change the meaning of the already existing values `"workspace::SendKeystrokes", "x y z",` which would be surprising, you'd need to go back and wrap them in `[]`, like `[["workspace::SendKeystrokes", "x y z"], ["vim::Right"]]`

> How does evil mode handle `a <escape>` do you end up moving forward?

Yes. `a` puts the cursor in insert mode one place ahead of where it is, so it make sense to me that it would stay there after you exit insert mode.

> adding a lot of conditional logic to emulate various different vim emulators

I'm only mentioning Evil mode because it's prior art for giving an option to fix what is just a mistake in vim's design, not because I want Zed to be compatible with an option in that emulator specifically.
